steel/src/history/beacon_roots.rs

```diff
@@ -170,16 +121,113 @@ impl BeaconRootsContract {
         };
 
         // validate the returned state and compute the return value
-        match Self::get_from_state(state.clone(), calldata) {
+        match BeaconRootsContract::get_from_db(&mut state, calldata) {
             Ok(returns) => Ok((returns, state)),
             Err(err) => match err {
                 Error::Reverted => Err(anyhow!("BeaconRootsContract({}) reverted", calldata)),
-                err => Err(err).context("API returned invalid state"),
+                err => Err(err).context("RPC error"),
             },
         }
     }
 }
 
+/// Implements the Database trait, but only for the account of the beacon roots contract.
+impl Database for BeaconRootsState {
+    type Error = Error;
+
+    #[inline(always)]
+    fn basic(&mut self, address: Address) -> Result<Option<AccountInfo>, Self::Error> {
+        // only allow accessing the beacon roots contract's address
+        assert_eq!(address, ADDRESS);
+        let account: StateAccount = self.state_trie.get_rlp(ADDRESS_HASH)?.unwrap_or_default();
+        // and the account storage must match the storage trie
+        if account.storage_root != self.storage_trie.hash_slow() {
+            return Err(Error::InvalidState);
+        }
+
+        Ok(Some(AccountInfo {
+            balance: account.balance,
+            nonce: account.nonce,
+            code_hash: account.code_hash,
+            code: None,
+        }))
+    }
+
+    fn code_by_hash(&mut self, _code_hash: B256) -> Result<Bytecode, Self::Error> {
+        unreachable!()
```

Comment:
Semantically, this is closer to being unimplemented
```suggestion
        // should never be called.
        unimplemented!()
```